Sporadic test failure: TestUserJoiningPopulatedChannel

In: http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/127/console

```
2017-11-15T11:32:31.519-08:00 Changes: simple changes cannot get Close Notifier from ResponseWriter
2017-11-15T11:32:31.519-08:00 Changes+: MultiChangesFeed: channels expand to "!:1,alpha:102" ...   (to user2)
2017-11-15T11:32:31.519-08:00 Cache: getCachedChanges("!", 0) --> 0 changes valid from #1
2017-11-15T11:32:31.519-08:00 Changes+: [changesFeed] Found 0 changes for channel !
2017-11-15T11:32:31.519-08:00 Changes+: Grant for channel [alpha] is after the current sequence - skipped for this iteration.  Grant:[102] Current:[101]   (to user2)
2017-11-15T11:32:31.519-08:00 Changes+: Found sequence later than stable sequence: stable:[101] entry:[102] (_user/user2)
2017-11-15T11:32:31.519-08:00 Changes: MultiChangesFeed done   (to user2)
2017-11-15 11:32:31.519743 I | _changes looks like: {"results":[
],
"last_seq":"0"}

	api_test.go:1944
	assert.Equals(t, len(changes.Results), 101)	

	expected: 101
	     got: 0
2017-11-15T11:32:31.520-08:00 Changes+: changeListener.Stop() called
--- FAIL: TestUserJoiningPopulatedChannel (5.68s)
	assert.go:31: 
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range
```

(from https://github.com/couchbase/sync_gateway/pull/3066 originally).

Also, was reported previously here too: https://github.com/couchbase/sync_gateway/issues/2612

--- a/rest/api_test.go
+++ b/rest/api_test.go
@@ -1939,29 +1939,20 @@ func TestUserJoiningPopulatedChannel(t *testing.T) {
 		assertStatus(t, rt.Send(request("PUT", docpath, `{"foo": "bar", "channels":["alpha"]}`)), 201)
 	}
 
-	// Check the _changes feed with limit, to split feed into two
-	var changes struct {
-		Results []db.ChangeEntry
-	}
-
 	limit := 50
-	response = rt.Send(requestByUser("GET", fmt.Sprintf("/db/_changes?limit=%d", limit), "", "user1"))
-	log.Printf("_changes looks like: %s", response.Body.Bytes())
-	err = json.Unmarshal(response.Body.Bytes(), &changes)
-	assert.Equals(t, err, nil)
-	assert.Equals(t, len(changes.Results), 50)
-	since := changes.Results[49].Seq
-	assert.Equals(t, changes.Results[49].ID, "doc48")
+	changesResults, err := rt.WaitForChanges(50, fmt.Sprintf("/db/_changes?limit=%d", limit), "user1")
+	assertNoError(t, err, "Unexpected error")
+	assert.Equals(t, len(changesResults.Results), 50)
+	since := changesResults.Results[49].Seq
+	assert.Equals(t, changesResults.Results[49].ID, "doc48")
 	assert.Equals(t, since.Seq, uint64(50))
 
 	//// Check the _changes feed with  since and limit, to get second half of feed
-	response = rt.Send(requestByUser("GET", fmt.Sprintf("/db/_changes?since=\"%s\"&limit=%d", since, limit), "", "user1"))
-	log.Printf("_changes looks like: %s", response.Body.Bytes())
-	err = json.Unmarshal(response.Body.Bytes(), &changes)
-	assert.Equals(t, err, nil)
-	assert.Equals(t, len(changes.Results), 50)
-	since = changes.Results[49].Seq
-	assert.Equals(t, changes.Results[49].ID, "doc98")
+	changesResults, err = rt.WaitForChanges(50, fmt.Sprintf("/db/_changes?since=\"%s\"&limit=%d", since, limit), "user1")
+	assertNoError(t, err, "Unexpected error")
+	assert.Equals(t, len(changesResults.Results), 50)
+	since = changesResults.Results[49].Seq
+	assert.Equals(t, changesResults.Results[49].ID, "doc98")
 	assert.Equals(t, since.Seq, uint64(100))
 
 	// Create user2
@@ -1969,57 +1960,47 @@ func TestUserJoiningPopulatedChannel(t *testing.T) {
 	assertStatus(t, response, 201)
 
 	//Retrieve all changes for user2 with no limits
-	response = rt.Send(requestByUser("GET", fmt.Sprintf("/db/_changes"), "", "user2"))
-	log.Printf("_changes looks like: %s", response.Body.Bytes())
-	err = json.Unmarshal(response.Body.Bytes(), &changes)
-	assert.Equals(t, err, nil)
-	assert.Equals(t, len(changes.Results), 101)
-	assert.Equals(t, changes.Results[99].ID, "doc99")
+	changesResults, err = rt.WaitForChanges(101, fmt.Sprintf("/db/_changes"), "user2")
+	assertNoError(t, err, "Unexpected error")
+	assert.Equals(t, len(changesResults.Results), 101)
+	assert.Equals(t, changesResults.Results[99].ID, "doc99")
 
 	// Create user3
 	response = rt.SendAdminRequest("PUT", "/db/_user/user3", `{"email":"user3@couchbase.com", "password":"letmein", "admin_channels":["alpha"]}`)
 	assertStatus(t, response, 201)
 
 	//Get first 50 document changes
-	response = rt.Send(requestByUser("GET", fmt.Sprintf("/db/_changes?limit=%d", limit), "", "user3"))
-	log.Printf("_changes looks like: %s", response.Body.Bytes())
-	err = json.Unmarshal(response.Body.Bytes(), &changes)
-	assert.Equals(t, err, nil)
-	assert.Equals(t, len(changes.Results), 50)
-	since = changes.Results[49].Seq
-	assert.Equals(t, changes.Results[49].ID, "doc49")
+	changesResults, err = rt.WaitForChanges(50, fmt.Sprintf("/db/_changes?limit=%d", limit), "user3")
+	assertNoError(t, err, "Unexpected error")
+	assert.Equals(t, len(changesResults.Results), 50)
+	since = changesResults.Results[49].Seq
+	assert.Equals(t, changesResults.Results[49].ID, "doc49")
 	assert.Equals(t, since.Seq, uint64(51))
 	assert.Equals(t, since.TriggeredBy, uint64(103))
 
 	//// Get remainder of changes i.e. no limit parameter
-	response = rt.Send(requestByUser("GET", fmt.Sprintf("/db/_changes?since=\"%s\"", since), "", "user3"))
-	log.Printf("_changes looks like: %s", response.Body.Bytes())
-	err = json.Unmarshal(response.Body.Bytes(), &changes)
-	assert.Equals(t, err, nil)
-	assert.Equals(t, len(changes.Results), 51)
-	assert.Equals(t, changes.Results[49].ID, "doc99")
+	changesResults, err = rt.WaitForChanges(51, fmt.Sprintf("/db/_changes?since=\"%s\"", since), "user3")
+	assertNoError(t, err, "Unexpected error")
+	assert.Equals(t, len(changesResults.Results), 51)
+	assert.Equals(t, changesResults.Results[49].ID, "doc99")
 
 	// Create user4
 	response = rt.SendAdminRequest("PUT", "/db/_user/user4", `{"email":"user4@couchbase.com", "password":"letmein", "admin_channels":["alpha"]}`)
 	assertStatus(t, response, 201)
 
-	response = rt.Send(requestByUser("GET", fmt.Sprintf("/db/_changes?limit=%d", limit), "", "user4"))
-	log.Printf("_changes looks like: %s", response.Body.Bytes())
-	err = json.Unmarshal(response.Body.Bytes(), &changes)
-	assert.Equals(t, err, nil)
-	assert.Equals(t, len(changes.Results), 50)
-	since = changes.Results[49].Seq
-	assert.Equals(t, changes.Results[49].ID, "doc49")
+	changesResults, err = rt.WaitForChanges(50, fmt.Sprintf("/db/_changes?limit=%d", limit), "user4")
+	assertNoError(t, err, "Unexpected error")
+	assert.Equals(t, len(changesResults.Results), 50)
+	since = changesResults.Results[49].Seq
+	assert.Equals(t, changesResults.Results[49].ID, "doc49")
 	assert.Equals(t, since.Seq, uint64(51))
 	assert.Equals(t, since.TriggeredBy, uint64(104))
 
 	//// Check the _changes feed with  since and limit, to get second half of feed
-	response = rt.Send(requestByUser("GET", fmt.Sprintf("/db/_changes?since=%s&limit=%d", since, limit), "", "user4"))
-	log.Printf("_changes looks like: %s", response.Body.Bytes())
-	err = json.Unmarshal(response.Body.Bytes(), &changes)
+	changesResults, err = rt.WaitForChanges(50, fmt.Sprintf("/db/_changes?since=%s&limit=%d", since, limit), "user4")
 	assert.Equals(t, err, nil)
-	assert.Equals(t, len(changes.Results), 50)
-	assert.Equals(t, changes.Results[49].ID, "doc99")
+	assert.Equals(t, len(changesResults.Results), 50)
+	assert.Equals(t, changesResults.Results[49].ID, "doc99")
 
 }
 

